base/base32768_body.hpp

```diff
@@ -17,22 +19,105 @@ namespace base {
 
 namespace internal_base32768 {
 
-constexpr std::int64_t bits_per_byte = 8;
-constexpr std::int64_t bits_per_code_point = 15;
-constexpr std::int64_t bits_per_final_code_point = 7;
-constexpr std::int64_t bytes_per_code_point =
-    (bits_per_code_point + 2 * bits_per_byte - 2) / bits_per_byte;
-static_assert(bytes_per_code_point == 3,
-              "End of input padding below won't be correct");
+// Ceiling log2 of n.  8 -> 3, 7 -> 2.
+constexpr int CeilingLog2(int const n) {
+  return n == 1 ? 0 : CeilingLog2(n >> 1) + 1;
+}
+
+// A repertoire is used to encode or decode integer into code points.
+class Repertoire {
+ public:
+  virtual char16_t const& Encode(std::uint16_t const k) const = 0;
+  virtual std::uint16_t const& Decode(char16_t const code_point) const = 0;
+};
 
-enum Repertoire {
-  Begin = 0,
-  FifteenBits = 0,
-  SevenBits = 1,
-  End = 2,
+// A caching repertoire builds at constructions caches for fast encoding and
+// decoding.
+template<std::int64_t block_size, std::int64_t block_count>
+class CachingRepertoire : public Repertoire {
+ public:
+  template<std::int64_t block_count_plus_1>
+  constexpr CachingRepertoire(
+      char16_t const (&blocks_begin)[block_count_plus_1]);
```

Comment:
Make this private and befriend the factory?

Reply:
Done.